pkg/tbtc/node.go

```diff
@@ -1,17 +1,222 @@
 	"github.com/keep-network/keep-core/pkg/generator"
 	"github.com/keep-network/keep-core/pkg/internal/testutils"
 	"github.com/keep-network/keep-core/pkg/net"
 	"github.com/keep-network/keep-core/pkg/protocol/group"
 	"github.com/keep-network/keep-core/pkg/tecdsa/dkg"
 )
 
+// dkgResultSigner is responsible for signing the DKG result and verification of
+//  signatures generated by other group members.
+type dkgResultSigner struct {
+	chain Chain
+}
+
+func newDkgResultSigner(chain Chain) *dkgResultSigner {
+	return &dkgResultSigner{
+		chain: chain,
+	}
+}
+
+// SignResult signs the provided DKG result. It returns the information
+// pertaining to the signing process: public key, signature, result hash.
+func (drs *dkgResultSigner) SignResult(result *dkg.Result) (*dkg.SignedResult, error) {
+	resultHash, err := drs.chain.CalculateDKGResultHash(result)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"dkg result hash calculation failed [%v]",
+			err,
+		)
+	}
+
+	signing := drs.chain.Signing()
+
+	signature, err := signing.Sign(resultHash[:])
+	if err != nil {
+		return nil, fmt.Errorf(
+			"dkg result hash signing failed [%v]",
+			err,
+		)
+	}
+
+	return &dkg.SignedResult{
+		PublicKey:  signing.PublicKey(),
+		Signature:  signature,
+		ResultHash: resultHash,
+	}, nil
+}
+
+// VerifySignature verifies if the signature was generated from the provided
+// DKG result has using the provided public key.
+func (drs *dkgResultSigner) VerifySignature(signedResult *dkg.SignedResult) (bool, error) {
+	return drs.chain.Signing().VerifyWithPublicKey(
+		signedResult.ResultHash[:],
+		signedResult.Signature,
+		signedResult.PublicKey,
+	)
+}
+
+// dkgResultSubmitter is responsible for submitting the DKG result to the chain.
+type dkgResultSubmitter struct {
+	chain Chain
+}
+
+func newDkgResultSubmitter(chain Chain) *dkgResultSubmitter {
+	return &dkgResultSubmitter{
+		chain: chain,
+	}
+}
+
+// SubmitResult submits the DKG result along with submitting signatures to the
+// chain. In the process, it checks if the number of signatures is above
+// the required threshold, whether the result was already submitted and waits
+// until the member is eligible for DKG result submission.
+func (drs *dkgResultSubmitter) SubmitResult(
+	result *dkg.Result,
+	signatures map[group.MemberIndex][]byte,
+	startBlockNumber uint64,
+	memberIndex group.MemberIndex,
+) error {
+	config := drs.chain.GetConfig()
+
+	// Chain rejects the result if it has less than 25% safety margin.
+	// If there are not enough signatures to preserve the margin, it does not
+	// make sense to submit the result.
+	signatureThreshold := config.HonestThreshold +
+		(config.GroupSize-config.HonestThreshold)/2
+	if len(signatures) < signatureThreshold {
+		return fmt.Errorf(
+			"could not submit result with [%v] signatures for signature threshold [%v]",
+			len(signatures),
+			signatureThreshold,
+		)
+	}
+
+	onSubmittedResultChan := make(chan uint64)
```

Comment:
s/onSubmittedResultChan/resultSubmittedChan

Reply:
Done in 8eab2eb3e1ab0a377c70180d4b1a8a5df419d953.